src/com/google/enterprise/adaptor/database/TupleReader.java

```diff
@@ -205,45 +289,53 @@ protected void emitRow(ResultSet resultSet, ContentHandler handler)
         case Types.CHAR:
         case Types.VARCHAR:
           String string = resultSet.getString(col);
-          handler.startElement("", columnName, columnName, atts);
-          outWriter.write(string);
-          break;
-        case Types.CLOB:
-        case Types.LONGVARCHAR:
-          Reader reader = resultSet.getCharacterStream(col);
-          handler.startElement("", columnName, columnName, atts);
-          copyValidXMLCharacters(reader, outWriter);
-          reader.close();
+          if (string != null) {
+            try (Writer writer =
+                new BufferedWriter(new ContentHandlerWriter(handler))) {
+              handler.startElement("", columnName, columnName, atts);
+              writer.write(string);
+            }
+          }
           break;
         case Types.NCHAR:
-        case Types.NCLOB:
         case Types.NVARCHAR:
         case Types.LONGNVARCHAR:
-          Reader nReader = resultSet.getNCharacterStream(col);
-          handler.startElement("", columnName, columnName, atts);
-          copyValidXMLCharacters(nReader, outWriter);
-          nReader.close();
+          try (Reader reader = resultSet.getNCharacterStream(col);
+              Writer writer =
+                  new BufferedWriter(new ContentHandlerWriter(handler))) {
+            if (reader != null) {
+              handler.startElement("", columnName, columnName, atts);
+              copyValidXMLCharacters(reader, writer);
```

Comment:
Another weirdness here, and maybe this is related to Lavanya's fix, is that we were using this copyValidXMLCharacters for streams, but we're not using it for strings. I think we should split out the *CHAR types, and use getString, but wrap them in a StringReader and call copyValidXMLCharacters. (We don't need it for the integer types, which can't have invalid XML characters to begin with.)